roofit/roofitcore/inc/RooMinimizer.h

```diff
@@ -100,37 +117,103 @@ class RooMinimizer : public TObject {
 
-  RooMinimizerFcn *_fcn;
-  std::string _minimizerType;
+  RooAbsMinimizerFcn *_fcn;
+  std::string _minimizerType = "Minuit2";
+  FcnMode _fcnMode;
 
   static ROOT::Fit::Fitter *_theFitter ;
 
   std::vector<std::pair<std::string,int> > _statusHistory ;
 
   RooMinimizer(const RooMinimizer&) ;
-	
+
   ClassDef(RooMinimizer,1) // RooFit interface to ROOT::Fit::Fitter
 } ;
 
+////////////////////////////////////////////////////////////////////////////////
+/// Construct MINUIT interface to given function. Function can be anything,
+/// but is typically a -log(likelihood) implemented by RooNLLVar or a chi^2
+/// (implemented by RooChi2Var). Other frequent use cases are a RooAddition
+/// of a RooNLLVar plus a penalty or constraint term. This class propagates
+/// all RooFit information (floating parameters, their values and errors)
+/// to MINUIT before each MINUIT call and propagates all MINUIT information
+/// back to the RooFit object at the end of each call (updated parameter
+/// values, their (asymmetric errors) etc. The default MINUIT error level
+/// for HESSE and MINOS error analysis is taken from the defaultErrorLevel()
+/// value of the input function.
+
+template <typename MinimizerFcn>
+RooMinimizer::RooMinimizer(RooAbsReal &function, MinimizerFcn* /* value unused */)
+{
+   RooSentinel::activate();
+
+   if (_theFitter)
+      delete _theFitter;
+   _theFitter = new ROOT::Fit::Fitter;
+   _theFitter->Config().SetMinimizer(_minimizerType.c_str());
+   setEps(1.0); // default tolerance
+
+   _fcn = new MinimizerFcn(&function, this, _verbose);
+
+   // make sure to order correctly so that child classes get checked for first TODO
+   if (dynamic_cast<RooGradMinimizerFcn *>(_fcn)) {
+      _fcnMode = FcnMode::gradient;
+   } else if (dynamic_cast<RooMinimizerFcn *>(_fcn)) {
+      _fcnMode = FcnMode::classic;
+   } else {
+      throw std::logic_error("RooMinimizer's MinimizerFcn template argument must be (a subclass of) RooMinimizerFcn or "
+                             "RooGradMinimizerFcn. MinuitFcnGrad is used in the other RooMinimizer ctor.");
+   }
+
+   // default max number of calls
+   _theFitter->Config().MinimizerOptions().SetMaxIterations(500 * _fcn->getNDim());
+   _theFitter->Config().MinimizerOptions().SetMaxFunctionCalls(500 * _fcn->getNDim());
+
+   // Shut up for now
+   setPrintLevel(-1);
+
+   // Use +0.5 for 1-sigma errors
+   setErrorLevel(function.defaultErrorLevel());
+
+   // Declare our parameters to MINUIT
+   _fcn->Synchronize(_theFitter->Config().ParamsSettings(), _fcn->getOptConst(), _verbose);
+
+   // Now set default verbosity
+   if (RooMsgService::instance().silentMode()) {
+      setPrintLevel(-1);
+   } else {
+      setPrintLevel(1);
+   }
+}
+
+// static function
+template <typename MinimizerFcn>
+std::unique_ptr<RooMinimizer> RooMinimizer::create(RooAbsReal &function) {
+   return std::unique_ptr<RooMinimizer>(new RooMinimizer(function, static_cast<MinimizerFcn*>(nullptr)));
```

Comment:
```suggestion
   return std::make_unique<RooMinimizer>(function, static_cast<MinimizerFcn*>(nullptr));
```
Since we use C++14 as the minimum standard now, we you can use `std::make_unique`.